### DISCUSSION TITLE
build py file to so for android

### DISCUSSION THREAD

**Discussion by Jzhenli** [beeware/briefcase#1997] [Q&A]
Has anyone ever tried to compile a .py file to a .so file for android?  if yes, would u share some experiences,thanks.

**Comment by mhsmith** [comment]
See https://github.com/chaquo/chaquopy/issues/810.

**Comment by Jzhenli** [comment]
@mhsmith thanks for your reply, from your link, seems cython will build one py file to one so file, if I want to build a python package (may include many python files in a folder) to a single so file,  how to do it?  I often use [Nuitka](https://nuitka.net/user-documentation/use-cases.html#use-case-3-package-compilation) to do the same thing, not sure how to do it with cython?

**Reply by mhsmith** [reply]
You could compile them into multiple .so files, or you can use Cython's `include` directive. See the Cython documentation for more details.

My advice in this area depends on why you want to do this compilation:

* If it's for performance, you're unlikely to get any significant improvement unless you also use Cython-specific features like type annotation.
* If it's for obfuscation, see the [Chaquopy FAQ](https://chaquo.com/chaquopy/doc/current/faq.html#faq-obfuscate).

**Reply by Jzhenli** [reply]
@mhsmith Just for obfuscation.  My code files always more than one hundred, and if compile all of them into multiple .so files will be too messy. Nuitka can compile the package into one  .so, and it can also hide the code structure. Let me check if cython have the feature.

**Comment by Jzhenli** [comment]
According to [nuitka docs](https://nuitka.net/user-documentation/user-manual.html), it support many platforms such as Windows, Linux, macOS, Android....,   and it can also build code folder to a single so or pyd file which is helpful for code obfuscation, I have test nuitka with briefcase on windows and linux,  the build app can works well.  

When we init a project with briefcase, we will get the code structure as follows,
<img width="289" alt="image" src="https://github.com/user-attachments/assets/8e6121cf-7e4e-43ba-9b9d-7f5c1d2cb007">

we put all our core code as a python package in the app folder, then we compile them with nutika to a single app.pyd or app.so, and then package with briefcase. The final output as follows
<img width="217" alt="image" src="https://github.com/user-attachments/assets/4674ef90-f8ec-491a-9ae5-817e2fe3a09f">

Now I'm not sure how to do it for the android app,  seems android have some special settings.

**Reply by freakboy3742** [reply]
If Nuitka claims to support Android, this would seem to be a question for Nuitka's support channels. We don't make any claim of Nuitka compatibility for Briefcase.
